src/util/simplify_expr_boolean.cpp

```diff
@@ -121,116 +127,105 @@ bool simplify_exprt::simplify_boolean(exprt &expr)
 
       if(erase)
       {
-        it=operands.erase(it);
+        it = new_operands.erase(it);
         no_change = false;
       }
       else
         it++;
     }
 
     // search for a and !a
-    for(const exprt &op : operands)
+    for(const exprt &op : new_operands)
       if(op.id()==ID_not &&
          op.operands().size()==1 &&
          op.type().id()==ID_bool &&
          expr_set.find(op.op0())!=expr_set.end())
       {
-        expr = make_boolean_expr(expr.id() == ID_or);
-        return false;
+        return make_boolean_expr(expr.id() == ID_or);
       }
 
-    if(operands.empty())
+    if(new_operands.empty())
     {
-      expr = make_boolean_expr(expr.id() == ID_and);
-      return false;
+      return make_boolean_expr(expr.id() == ID_and);
     }
-    else if(operands.size()==1)
+    else if(new_operands.size() == 1)
     {
-      exprt tmp(operands.front());
-      expr.swap(tmp);
-      return false;
+      return std::move(new_operands.front());
     }
 
-    return no_change;
+    if(!no_change)
+    {
+      auto tmp = expr;
+      tmp.operands() = std::move(new_operands);
+      return std::move(tmp);
+    }
   }
 
-  return true;
+  return unchanged(expr);
 }
 
-bool simplify_exprt::simplify_not(exprt &expr)
+simplify_exprt::resultt<> simplify_exprt::simplify_not(const exprt &expr)
 {
   if(expr.operands().size()!=1)
-    return true;
+    return unchanged(expr);
 
-  exprt &op=expr.op0();
+  const exprt &op = expr.op0();
 
   if(expr.type().id()!=ID_bool ||
      op.type().id()!=ID_bool)
-    return true;
+  {
+    return unchanged(expr);
+  }
```

Comment:
Should be covered by `to_not_expr` - not sure whether it currently is, but I believe it should be done there.